### ISSUE TITLE
Support the .NET Standard Library

### ISSUE THREAD

**Issue by giggio** [mmanela/diffplex#12]
Please support the .NET Standard Library: https://docs.microsoft.com/en-us/dotnet/articles/standard/library

**Comment by mmanela** [comment]
What is needed to support DNX?

**Comment by giggio** [comment]
You would need to migrate from a .csproj/package.config project file structure to a .xproj/project.json file structure. With it, you don't need the nuspec files anymore, as project.json already does all that work. It is a considerable change, so you might want to check that more carefully.
To be fair, current 4.0 should already work with dnx. So, a better title would be "support dnx with .NET core". I will update that.

**Comment by mmanela** [comment]
If I am already a "portable" library does that not mean I work on core? Do I need the DNX project structure to target core?

**Comment by giggio** [comment]
No, a PCL will not work with .NET Core. You have to specifically target it.
If you use VS 2015 and try creating a `Console Application (Package)` you will see `dnxcore` as a runtime. You will see that you will not be able to add a reference to `diffplex`.

**Comment by mmanela** [comment]
It doesn't seem to list dnxcore here:

![image](https://cloud.githubusercontent.com/assets/304410/11104988/f0d34876-8880-11e5-8e81-64261ccdf79b.png)


What do I need to do?

**Comment by giggio** [comment]
You are creating the incorrect project type. The dnx projects are still under the Web node, on the project creation tool. Look for the `Console Application (Package)` project on VS 2015.

**Comment by mmanela** [comment]
Ok, I can take a look at that.

**Comment by giggio** [comment]
This is what will happen if you try to target `dnxcore`:

![diffplex](https://cloud.githubusercontent.com/assets/334958/11105283/154d122a-88b5-11e5-8ae2-c44d0d0e0506.png)

Here is the `project.json` file contents:

````json
{
  "version": "1.0.0-*",
  "description": "ConsoleApp2 Console Application",
  "authors": [ "giggio" ],
  "tags": [ "" ],
  "projectUrl": "",
  "licenseUrl": "",

  "dependencies": {
      "diffplex": "1.2.1"
  },

  "commands": {
    "ConsoleApp2": "ConsoleApp2"
  },

  "frameworks": {
    "dnx451": { },
    "dnxcore50": {
      "dependencies": {
        "Microsoft.CSharp": "4.0.1-beta-23409",
        "System.Collections": "4.0.11-beta-23409",
        "System.Console": "4.0.0-beta-23409",
        "System.Linq": "4.0.1-beta-23409",
        "System.Threading": "4.0.11-beta-23409"
      }
    }
  }
}
````

If you remove `dnxcore50` it compiles and runs just fine.

**Comment by mmanela** [comment]
If I build diffplex for .net core can someone use it from .net 4.5 ?

**Comment by giggio** [comment]
Hm... That is a good question. I would say that probably not, at this moment.
If you checkout this document: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/standard-platform.md
You will see how these things will work together.
If you decide to target the new `netstandard`, everything would just work. But that is not done yet.

**Comment by nguyenm100** [comment]
I believe the above link references changes for rc2 (currently things are at rc1 with rc2 slated for 2/2016).  In RC1, I believe you want to mark your (shared) assemblies as "donet5.4" (this changes to netstandard in RC2) to be completely neutral (and anyone can use it) and mark your app dnxcore50 to use dnx instead of (windows).net 4.6

The naming is horribly confusing :-(

**Comment by giggio** [comment]
Yes, the new .NET Standard will allow you to support .NET Core and .NET 4.5.

**Comment by russcam** [comment]
I have [a working DNX build](https://github.com/russcam/diffplex/tree/dnx-support) - once I've run the tests, I'll submit a PR :+1:

**Comment by Daniel15** [comment]
Any updates here?

**Comment by russcam** [comment]
I got sidetracked on other things; I'll update the DNX version I have to use the new tooling and update the PR

**Comment by giggio** [comment]
I updated the description, we don't need DNX support anymore, we need .NET Standard Support. :)

**Comment by giggio** [comment]
@russcam when are you planning on updating this PR?

**Comment by mmanela** [comment]
I took a stab at changing diffplex into a netstandard dll but I am hitting issues. Would love someone who has more experience with this to help. I pushed my changes here: https://github.com/mmanela/diffplex/tree/dotnetcore

I attempted to convert DiffPlex, DiffPlex.ConsoleRunner, Facts.DiffPlex and Perf.DiffPlex to netstandard. I omitted the other two project for now to keep it simple. However, I am getting errors like:

![image](https://cloud.githubusercontent.com/assets/304410/20459815/65fac1cc-ae84-11e6-9b60-d8806f22436a.png)

@russcam @giggio @AArnott Could any of you help?

**Comment by russcam** [comment]
I got around to updating the PR #14 that I had for updating to DNX and have ported it over to using .NET Core 1.0 in #17. Additionally, I have added a Build project that uses [FAKE](http://fsharp.github.io/FAKE/index.html) to build the solution and run unit tests. Because FAKE is pulled in through nuget, it means a nuget restore needs to be done before the `build.bat` build script can run. 

If interested, I can port nuget over to using [paket](https://fsprojects.github.io/Paket/) and check in only the paket,bootstrapper.exe into the repository.

**Comment by mmanela** [comment]
@russcam With 2017 RC they no longer use the json files and its now using a simple version of project xml. In my branch (https://github.com/mmanela/diffplex/tree/dotnetcore) it looked pretty simple although I hit that odd dependency issue. I think we should move to this since this is where its going to be with the new release of VS.

**Comment by russcam** [comment]
@mmanela, ok. I'll need to install VS 2017 RC and take a look

**Comment by mmanela** [comment]
Good news, I got it working in my branch https://github.com/mmanela/diffplex/tree/dotnetcore

**Comment by mmanela** [comment]
Just published a new package that targets netstandard 1.4
